### PR TITLE
[DO NOT MERGE] - iOS: Integrate MMB into RNNativePlayground

### PR DESCRIPTION
**Pull request by lukewalczak**
- [x] - refactor to apply more RN Apps
- [x] - integrate MMB
- [x] - prepare native `Log in` to get `AccessToken` and pass it to `Bookings`
- [x] - fix navigation to work properly
- [x] - prepare standalone package for `MMB`

### CODE REVIEW

File: app/core/src/screens/MMBPackageWrapper.js [hunk 1]
```diff
@@ -42,7 +47,8 @@ export default class MMBPackageWrapper extends React.Component<
 
   componentDidMount = () => {
     this.fetchToken();
-    this.props.navigation.setParams({ fetchToken: this.fetchToken });
+    this.props.navigation &&
+      this.props.navigation.setParams({ fetchToken: this.fetchToken });
```
Comment (grabbou) [inline app/core/src/screens/MMBPackageWrapper.js]: Let's clear this file and create an appropriate "StandalonePackage" file

Comment (grabbou) [reply app/core/src/screens/MMBPackageWrapper.js]: Should be `appRegistry` folder with root component and standalone packages.

Comment (tbergq) [reply app/core/src/screens/MMBPackageWrapper.js]: this file is not supposed to be a part of the native application. This is just a rapper to pass in to `ManageMyBookingPackage` what native should pass into the application

File: app/MMB/src/appRegistry/MMBStandalonePackage.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import NavigationStack from '../../src/navigation/NavigationStack';
+import RootComponent from './RootComponent';
+
+type Props = {|
+  dimensions: DimensionType,
+  onNavigationStateChange: () => void,
+  accessToken: string,
+  currency: string,
+|};
+
```
Comment (tbergq) [inline app/MMB/src/appRegistry/MMBStandalonePackage.js]: please use read-only prop types

File: app/MMB/src/appRegistry/MMBStandalonePackage.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  };
+
+  componentDidMount = () => {
+    this.fetchToken();
+  };
+
+  fetchToken = async () => {
+    const accessToken = await AsyncStorage.getItem('mobile:MMB-Token');
+    this.setState({ accessToken });
+  };
```
Comment (tbergq) [inline app/MMB/src/appRegistry/MMBStandalonePackage.js]: The token should be passed in as a prop from native code

File: app/MMB/src/appRegistry/MMBStandalonePackage.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+  renderInnerComponent = () => {
+    if (!this.state.accessToken && !this.props.accessToken) {
+      return (
+        <View style={styles.loginMessageContainer}>
+          <Translation passThrough="You are not logged in, go to profile and log in" />
+        </View>
+      );
+    }
```
Comment (tbergq) [inline app/MMB/src/appRegistry/MMBStandalonePackage.js]: This will be the responsibility of the native app to show this message.

File: app/MMB/src/appRegistry/MMBStandalonePackage.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    );
+  };
+
+  render = () => {
+    return (
+      <RootComponent
+        dimensions={this.props.dimensions}
+        accessToken={this.state.accessToken || this.props.accessToken}
+      >
+        {this.renderInnerComponent()}
```
Comment (tbergq) [inline app/MMB/src/appRegistry/MMBStandalonePackage.js]: And then this would just be 
```
 <NavigationStack
   onNavigationStateChange={this.props.onNavigationStateChange}
 />
```

File: app/MMB/src/appRegistry/RootComponent.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import { AuthContext } from '@kiwicom/mobile-relay';
+import { Dimensions, type DimensionType } from '@kiwicom/mobile-shared';
+
+import BookingDetailContext from '../context/BookingDetailContext';
+
+type Props = {|
+  children: React.Node,
+  dimensions: DimensionType,
+  accessToken: string,
+|};
```
Comment (tbergq) [inline app/MMB/src/appRegistry/RootComponent.js]: read-only props please

File: packages/relay/src/AuthContext.js [hunk 7]
```diff
@@ -37,6 +37,13 @@ class Provider extends React.Component<Props, State> {
     };
   }
 
+  static getDerivedStateFromProps(props: Props, state: State) {
+    return {
+      ...state,
+      accessToken: props.accessToken,
+    };
+  }
```
Comment (tbergq) [inline packages/relay/src/AuthContext.js]: Why is this needed, will the prop being passed from native ever change?

File: packages/shared/src/WithNativeNavigation.js [hunk 8]
```diff
@@ -9,7 +9,7 @@ type NavigationState = {|
 
 type InjectedProps = {|
   onNavigationStateChange: () => void,
-  onBackClicked: () => void,
+  onBackClicked?: () => void,
```
Comment (ferrannp) [inline packages/shared/src/WithNativeNavigation.js]: Why is this optional now? What has changed?

**Comment by ferrannp** [comment]
I see this code:

```objc
- (NSURL *)jsCodeLocation {
  // TODO: Change with hotels bundle
  return [NSURL URLWithString:@"http://localhost:8081/index.bundle?platform=ios&dev=true"];
}

- (NSURL *)jsCodeLocation {
  // TODO: Change with bookings bundle
  return [NSURL URLWithString:@"http://localhost:8081/index.bundle?platform=ios&dev=true"];
}
```

Does this mean we want to have two JS bundles instead of one ? Will we init two bridges? We will have two copies of React Native in each JS bundle? Even if we have two difference JS, we are adding all native dependencies (even the ones hotels do not need). How we handle that?

I am trying to thing if having more than one JS bundle is really the way to go. WDYT @grabbou ?

File: packages/shared/src/WithNativeNavigation.js (hunk 8 above)
Comment (lukewalczak) [reply packages/shared/src/WithNativeNavigation.js]: Reverted. Added `FlowFixMe` till `onBackClicked` will not be implmented in Android app.

**Comment by grabbou** [comment]
I think this is a question that Kiwi team has to answer themselves. I would myself go with a single bundle, because the additional KBs of Javascript are going to vanish when combined with native code along the way. One important trade-off of multiple bundles to understand is that separate bridges mean separate memory. That makes it hard to maintain user session or any other in-memory kind of data between multiple bridges. Keeping things simple usually delivers best performance.

We can always keep a single bundle and in the future, think about code splitting.

**Comment by ferrannp** [comment]
> If this feature is in the release bundle during it's development, it will increase the final APK size which is quite sensitive parameter for Android users

Yeah I agree on this. I would also not merge right now PR that adds extra JS code and extra native dependencies to the release apk if the code is not used.

Anyway this PR is for iOS and we decided to keep it on hold till the apps need to use MMB feature.

On the other hand, I also agree that branching/QA here would be useful. Just create the version library for Android or the framework for iOS and test it in the real app though (not our pure RN example). CC @tbergq

**Comment by tbergq** [comment]
> Anyway this PR is for iOS and we decided to keep it on hold till the apps need to use MMB feature.

So this means no increased size until it is released. 

For now, I see no need for branching since this workflow just creates more work. Maybe some time in the future it will make sense to change the workflow.

**Comment by lukewalczak** [comment]
Closed (obsolete): The structure was changed along with the PR #1083.
